src/dmd/func.d

```diff
@@ -2167,6 +2167,78 @@ extern (C++) class FuncDeclaration : Declaration
         return sf;
     }
 
+    /****************************************************
+     * Merge into this function the 'in' contracts of all it overrides.
+     */
+    extern (D) final Statement mergeFrequireInclusivePreview(Statement sf, Expressions* params)
+    {
+        /* If a base function and its override both have an IN contract, then
+         * the override in contract must widen the guarantee of the base contract.
+         * This is checked by generating:
+         *
+         * void derived.in() {
+         *  try {
+         *    ... body of derived.in() ...
+         *  }
+         *  catch () {
+         *    // derived in rejected this argument. so parent must also reject it, or we've tightened the contract.
+         *    base.in();
+         *    assert(false, "Logic error: " ~ thr.msg);
+         *  }
+         */
+
+        foreach (fdv; foverrides)
+        {
+            /* The semantic pass on the contracts of the overridden functions must
+             * be completed before code generation occurs.
+             * https://issues.dlang.org/show_bug.cgi?id=3602
+             */
+            if (fdv.frequires && fdv.semanticRun != PASS.semantic3done)
+            {
+                assert(fdv._scope);
+                Scope* sc = fdv._scope.push();
+                sc.stc &= ~STC.override_;
+                fdv.semantic3(sc);
```

Comment:
I just copypasted this from the original `mergeFrequire`. I don't actually know much about DMD, so I have to trust it works sensibly there.

Reply:
I see. Could you coalesce both functions s.t. we don't have to maintain two copies?

Reply:
I specifically duplicated the function because otherwise the control flow for this function, which is already a bit cryptic, becomes totally unreadable.